### ISSUE TITLE
Dangling pointer lint?

### ISSUE THREAD

**Issue by 1Dragoon** [rust-lang/rust-clippy#8015]
I'm not entirely sure whether unsafe rust is in-scope for clippy, but if it is, I'd like to see a feature to detect (at least) the more obvious cases of a dangling pointer. Specifically, I had trouble figuring out what was causing my code to have bugs here:

https://github.com/microsoft/windows-rs/issues/1342

This is the offending bit:
```rust
let pwfilter = PWSTR(format!("{}\0", filter).encode_utf16().collect::<Vec<u16>>().as_mut_ptr());
```

I'm still new to programming in general, especially rust, so I didn't really see this as a problem as I'm also used to the borrow checker working even in unsafe (though, as I've been taught, this isn't a borrow.)

**Comment by xFrednet** [comment]
Hey, thanks for the suggestion! Clippy has some lints targeted towards unsafe code, this lint sounds reasonable. Could you try to clarify the example a bit, I'm not sure where the dangling pointer comes from.

```rs
let _new = String::from("Test").encode_utf16().collect::<Vec<u16>>().as_mut_ptr();
```

Compiles fine on Rust's playground :)

**Comment by MarijnS95** [comment]
@xFrednet that example compiles fine, but the `Vec` goes out of scope after that statement meaning `_new` is now a raw pointer to freed memory. Using/dereferencing it is UB and for that reason already `unsafe`, but that brings us back to the age-old discussion of discarding lifetime information (all the `as_*ptr` variants) needing to be `unsafe` in the first place.

(Note: the underlying issue is that `windows-rs` is still working on safer bindings around string construction, without requiring raw pointers - or at least making those unsafe)

**Comment by xFrednet** [comment]
I haven't worked with unsafe code that much, so please correct my if I'm wrong but does the `Vec` really go out of scope? I expected that as first as well, but the line compiles fine in safe code. Only writing to it afterwards requires unsafe code. Could you maybe provide a minimal Rust example?

My current understanding is that this lint should check for all instances where `.as_mut_ptr()` or something similar is used on a value that *dies* before the pointer value does. Is that correct? :upside_down_face:

**Comment by MarijnS95** [comment]
> My current understanding is that this lint should check for all instances where `.as_mut_ptr()` or something similar is used on a value that dies before the pointer value does. Is that correct?

Exactly. It's actually a backwards thing to do because that's why we have borrows with lifetimes in the first place. In most cases a borrow coerces into a raw pointer nicely, but that's not the case with slices (ie. `&mut [u16]` doesn't coerce into `*mut u16`, though a `&myslice[0]` may help out here).

And even then, in this example for https://github.com/microsoft/windows-rs/issues/1342, it's probably hard(er) to track the raw pointer _through_ this `PWSTR` newtype constructor (as mentioned above, afaik this allows safe construction and `fmt::Debug` printing, so the creation of `PWSTR` should have been unsafe in the first place).

> I haven't worked with unsafe code that much, so please correct my if I'm wrong but does the `Vec` really go out of scope? I expected that as first as well, but the line compiles fine in safe code. Only writing to it afterwards requires unsafe code. Could you maybe provide a minimal Rust example?

Yes, it goes out of scope. Using a borrow (which is effectively a pointer _with_ lifetime information) and subsequently using that borrow will tell you:

```rust
fn main() {
    let new = String::from("Test").encode_utf16().collect::<Vec<u16>>().as_slice();
    dbg!(new);
}
```

```rust
error[E0716]: temporary value dropped while borrowed
 --> src/main.rs:2:15
  |
2 |     let new = String::from("Test").encode_utf16().collect::<Vec<u16>>().as_slice();
  |               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^           - temporary value is freed at the end of this statement
  |               |
  |               creates a temporary which is freed while still in use
3 |     dbg!(new);
  |          --- borrow later used here
  |
  = note: consider using a `let` binding to create a longer lived value

For more information about this error, try `rustc --explain E0716`.
```

> Only writing to it afterwards requires unsafe code

Fwiw _reading and_ writing is `unsafe` and UB here.

**Comment by xFrednet** [comment]
Alright, thank you for the clarification, that helped a lot! The lint implementation will require some tinkering, but would be an interesting project. I'm happy to provide guidance if someone wants to try their hand on this one :upside_down_face: 

---

@1Dragoon I think we can keep this issue as the lint suggestion. Would you mind updating the issue description to have "What it does" headline explaining that it checks for dangling pointers and when these occur? As well as an example and how it should be written, similar to the given example. That would be very helpful! Also, feel free to suggest new lint ideas if you have any, it's nice to hear that Clippy is a helping :)

**Comment by 1Dragoon** [comment]
Sorry was away for a while, just updated it

**Comment by y21** [comment]
rustc now has the `dangling_pointers_from_temporaries` lint that emits warnings for the repro case, so this can probably be closed?
```
   Compiling playground v0.0.1 (/playground)
warning: a dangling pointer will be produced because the temporary `Vec<u16>` will be dropped
 --> src/main.rs:2:74
  |
2 |     let _new = String::from("Test").encode_utf16().collect::<Vec<u16>>().as_mut_ptr();
  |                --------------------------------------------------------- ^^^^^^^^^^ this pointer will immediately be invalid
  |                |
  |                this `Vec<u16>` is deallocated at the end of the statement, bind it to a variable to extend its lifetime
  |
  = note: pointers do not have a lifetime; when calling `as_mut_ptr` the `Vec<u16>` will be deallocated at the end of the statement because nothing is referencing it as far as the type system is concerned
  = help: for more information, see <https://doc.rust-lang.org/reference/destructors.html>
  = note: `#[warn(dangling_pointers_from_temporaries)]` on by default

```
